src/gitlab.js

```diff
@@ -3,82 +3,127 @@
 // @flow
 
 import Gitlab from 'gitlab';
+import { AuthCache, UserData } from './authcache';
 import httperror from 'http-errors';
 import type { PackageAccess, Config, Logger, Callback } from '@verdaccio/types';
 
+export type VerdaccioGitlabConfig = Config & {
+  url: string,
+  authCache?: {
+    enabled?: boolean,
+    ttl?: number
+  }
+};
+
+export type VerdaccioGitlabStuff = {
+  [config: string]: {
+    users_file: string,
+    self_path: string
+  },
+  logger: Logger
+};
+
+type VerdaccioGitlabPackageAccess = PackageAccess & {
+  name: string,
+  gitlab?: boolean
+}
+
 export default class VerdaccioGitLab {
```

Comment:
I just shipped `IBasicAuth` interface to help with class structure.
https://github.com/verdaccio/flow-types/blob/master/src/types.js#L345

Reply:
Ok, but as I understand that commit, that is a non-backwards compatible change, isn't it? It means we have to build the plugin on top of the latest verdaccio version, ?

Reply:
this project already uses `"@verdaccio/types": "^3.0.0"` so it should works. I hope that's what you are referring for.

Reply:
Mmm ok maybe I'm not understanding this. In the current code I'm using, `allow_access` and `allow_publish` have the following signatures:

`allow_access(user: { name: string }, _package: VerdaccioGitlabPackageAccess, cb: Callback)`  
`allow_publish(user: { name: string, real_groups: string[] }, _package: VerdaccioGitlabPackageAccess, cb: Callback)`

But in the `@verdaccio/types` you linked, I see this in `verdaccio$IPluginAuth`:

`allow_access(packageName: string, user: string, cb: verdaccio$Callback): void;`  
`allow_publish(packageName: string, user: string, cb: verdaccio$Callback): void;`

Particularly, in the current code I receive an object structure for both the user and package data, whereas in the new interface i just get a string.

Am I missing anything?

Reply:
`@verdaccio/types@3.3.3` should be right alredy

Reply:
Just checked, in my code the user is still an object with properties, but the signature in the types just gives me a string, :-?

Reply:
When I try to implement the interface such as:

```
export default class VerdaccioGitLab implements IPluginAuth {
...
allow_access(user: { name: string }, _package: PackageAccess, cb: Callback) {
...
allow_publish(user: { name: string, real_groups: string[] }, _package: PackageAccess, cb: Callback) {
```

I have the following issues (apart from the mentioned `user: string` that is actually an object):

```
- Property gitlab is missing in verdaccio$PackageAccess [1].
- Cannot get _package.name because property name is missing in verdaccio$PackageAccess [1].
```

Any ideas? I'm actually using those fields in the provided objects, and they do exist. How to reconcile the type definition with what is actually needed?

Reply:
> Just checked, in my code the user is still an object with properties, but the signature in the types just gives me a string, :-?

omg ... sorry ... I must be asleep. I'll fix it again.

Reply:
` @verdaccio/types@3.4.1` add a `RemoteUser` and allow SubTypes for `PackageAccess` on  type that looks like

```
declare type verdaccio$RemoteUser = {
  real_groups: Array<string>;
  groups: Array<string>;
  name: string | void;
  error?: string;
}

declare interface verdaccio$IPluginAuth extends verdaccio$IPlugin {
  authenticate(user: string, password: string, cb: verdaccio$Callback): void;
  adduser(user: string, password: string, cb: verdaccio$Callback): void;
  allow_access(user: verdaccio$RemoteUser, pkg: $Subtype<verdaccio$PackageAccess>, cb: verdaccio$Callback): void;
  allow_publish(user: verdaccio$RemoteUser, pkg: $Subtype<verdaccio$PackageAccess>, cb: verdaccio$Callback): void;
}
```

Reply:
eg: https://github.com/verdaccio/verdaccio/pull/815/commits/cde695a18f444199e8d8e0d6174c3529d366aa39#diff-ac171c117d06952be3c28d8f4fbb871aR64